### PR TITLE
Samples rejection view

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

This PR adds a Samples rejection view where the user can choose the reasons of rejection and whether a notification email has to be sent to the client contact.

Linked issue: https://github.com/senaite/senaite.core/issues/1238

## Current behavior before PR

User cannot neither choose the reasons of rejection nor e-mail notification.

## Desired behavior after PR is merged

User can choose the reasons of rejection and e-mail notification.

Captura de 2019-03-28 19-19-37

Captura de 2019-03-28 19-17-47

Captura de 2019-03-28 20-11-08

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/browser/analysisrequest/reject_samples.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    """View that renders the Samples rejection view
+    """
+    template = ViewPageTemplateFile("templates/reject_samples.pt")
+
+    def __init__(self, context, request):
+        super(RejectSamplesView, self).__init__(context, request)
+        self.context = context
+        self.request = request
+        self.back_url = self.context.absolute_url()
+        self.notify = api.get_setup().getNotifyOnSampleRejection()
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/reject_samples.py]: Avoid doing such lookups in `__init__`. Better do something like this:

```python
@property
def notify(self):
    return api.get_setup().getNotifyOnSampleRejection()
```
maybe also view-cache it if neccessary

File: bika/lims/browser/analysisrequest/reject_samples.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+        # Form submit toggle
+        form_submitted = form.get("submitted", False)
+
+        # Buttons
+        form_continue = form.get("button_continue", False)
+        form_cancel = form.get("button_cancel", False)
+
+        # Is Rejection Workflow Enabled
+        if not api.get_setup().isRejectionWorkflowEnabled():
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/reject_samples.py]: more aesthetically, but you could do a property or method for this as well, maybe:

```
def is_rejection_workflow_enabled(self):
     return api.get_setup().isRejectionWorkflowEnabled()
```

File: bika/lims/browser/analysisrequest/reject_samples.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        for obj in self.get_samples_from_request():
+            obj = api.get_object(obj)
+            yield {
+                "obj": obj,
+                "id": api.get_id(obj),
+                "uid": api.get_uid(obj),
+                "title": api.get_title(obj),
+                "path": api.get_path(obj),
+                "url": api.get_url(obj),
+                "sample_type": api.get_title(obj.getSampleType()),
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/reject_samples.py]: I think we have `getSampleTypeTitle()` as well

File: bika/lims/browser/analysisrequest/reject_samples.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    def get_rejection_reasons(self):
+        """Returns the list of available rejection reasons
+        """
+        return api.get_setup().getRejectionReasonsItems()
+
+    def get_samples_data(self):
+        """Returns a list of Samples data (dictionary)
+        """
+        for obj in self.get_samples_from_request():
+            obj = api.get_object(obj)
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/reject_samples.py]: Trust your code, you know that you get objects here

File: bika/lims/browser/analysisrequest/templates/reject_samples.pt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                     type="submit"
+                     name="button_cancel"
+                     i18n:attributes="value"
+                     value="Cancel ⎋"/>
+              <!-- Continue -->
+              <input class="btn btn-success btn-sm"
+                     type="submit"
+                     name="button_continue"
+                     i18n:attributes="value"
+                     value="Reject ⇶"/>
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/templates/reject_samples.pt]: `i18ndude` reports errors with these unicode characters, better remove them

File: bika/lims/content/bikasetup.py [hunk 6]
```diff
@@ -916,6 +904,16 @@ def isRejectionWorkflowEnabled(self):
         else:
             return False
 
+    def getRejectionReasonsItems(self):
+        """Return the list of predefined rejection reasons
+        """
+        reasons = self.getRejectionReasons()
```
Comment (ramonski) [inline bika/lims/content/bikasetup.py]: I see that you removed the fied `RejectionReasons` above. Where does no this `getRejectionReasons ` come from?

Comment (xispa) [reply bika/lims/content/bikasetup.py]: There were two "RejectionReasons" schema fields:
https://github.com/senaite/senaite.core/pull/1309/files/0462fb2292c3f7d42a4e5c09ccd7c0cc18ae1c50#diff-0dde5125b635d6fb55f5372798dae571R801

https://github.com/senaite/senaite.core/pull/1309/files/0462fb2292c3f7d42a4e5c09ccd7c0cc18ae1c50#diff-0dde5125b635d6fb55f5372798dae571L561

I've removed one of them

**Pull request merged**